File: packages/laterite/python/laterite/ags_types.py
"""laterite.ags_types — Rust-backed AGS4 type system.

Exposes `parse_value` / `canonical_type` / `display_hint` semantics
backed by the Rust engine in `rust-packages/laterite-ags4-core/src/ags_types.rs`.

The `CanonicalType` StrEnum lives Python-side so existing
`canonical_type(x) is CanonicalType.DECIMAL` checks keep their
identity. Rust returns the lowercase label; we coerce back to the
enum.

`parse_value` returns native Python types — int / float / bool /
datetime.datetime / datetime.date / datetime.time / str / None —
matching the pure-Python implementation. Returns the trimmed string
for unknown AGS codes (passthrough), matching the prior behaviour.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Any

from . import _laterite_native as _native


class CanonicalType(StrEnum):
    """Cross-system target types. Values match the lowercase labels the
    Rust side returns from `canonical_type` so the enum doubles as the
    string interchange representation (e.g. for ``_spec_headings``)."""

    STRING = "string"
    INTEGER = "integer"
    DECIMAL = "decimal"
    DATETIME = "datetime"
    DATE = "date"
    TIME = "time"
    BOOL = "bool"
    ENUM = "enum"


def canonical_type(ags_type: str) -> CanonicalType:
    """AGS spec type code → canonical category.

    Raises ``ValueError`` for unknown codes — the AGS4 codec coerces
    passthrough groups' heading types to ``'X'`` before calling, so
    in practice this never raises on dictionary-resident headings.
    """
    label = _native.canonical_type(ags_type)
    if label is None:
        raise ValueError(f"unknown AGS type code: {ags_type!r}")
    return CanonicalType(label)


def display_hint(ags_type: str) -> str | None:
    """Presentation hint for a numeric AGS type, or ``None``.

    ``'2DP'`` → ``'%.2f'``, ``'3SF'`` → ``'%.3g'``, ``'1SCI'`` →
    ``'%.1e'``. String / datetime / bool types return ``None``.
    """
    return _native.display_hint(ags_type)


def parse_value(raw: Any, ags_type: str) -> Any:
    """Parse an AGS4-shaped raw string into the canonical Python type.

    Permissive: unparseable values return ``None``. The single source
    of truth for AGS4 + ``.agsx`` ingest (both Rust paths); the Python
    ``laterite_ags5x._codec`` module is encode-only since Stage E3.
    """
    # Accept any input PyO3 won't coerce directly (e.g. an int passed
    # by a careless caller) — match the Python implementation's
    # "isinstance(s, str)" guards by stringifying first.
    if raw is None:
        return None
    if not isinstance(raw, str):
        raw = str(raw)
    return _native.parse_value(raw, ags_type)


__all__ = [
    "CanonicalType",
    "canonical_type",
    "display_hint",
    "parse_value",
]

File: rust-packages/laterite-node/ts/ags-types.ts
// laterite.agsTypes — the AGS4 type system, the Node port of laterite-py's
// `ags_types`. The parsing LOGIC is native (one shared engine across hosts);
// this is just the typed TS face over it.
import { Ags4Error } from "./errors";
import {
  canonicalType as nativeCanonicalType,
  displayHint,
  parseValue as nativeParseValue,
} from "./native";

/** Cross-system target categories — the lowercase labels the engine returns. */
export type CanonicalType =
  | "string"
  | "integer"
  | "decimal"
  | "datetime"
  | "date"
  | "time"
  | "bool"
  | "enum";

/** A parsed AGS value: number (integer/decimal), boolean (YN), string
 * (text/enum **and** the canonical datetime/date/time strings), or null. */
export type AgsValue = string | number | boolean | null;

/**
 * AGS spec type code → canonical category. Throws for unknown codes (mirrors
 * Python's `ValueError`), so the engine's permissive `null` never leaks into
 * caller code as a silent miss.
 *
 * @param agsType - An AGS4 spec type code (e.g. `"2DP"`, `"ID"`, `"YN"`, `"DT"`).
 * @returns The lowercase canonical category the engine maps that code to.
 * @throws {Ags4Error} If the code is not a recognised AGS type.
 */
export function canonicalType(agsType: string): CanonicalType {
  const label = nativeCanonicalType(agsType);
  if (label === null)
    throw new Ags4Error(`unknown AGS type code: ${JSON.stringify(agsType)}`);
  return label as CanonicalType;
}

export { displayHint };

/** Coerce a non-null scalar to a string without the `[object Object]` footgun
 *  (an unexpected object → JSON; AGS cells are primitives, so that branch is
 *  defensive only). */
function scalarString(value: unknown): string {
  if (typeof value === "string") return value;
  if (
    typeof value === "number" ||
    typeof value === "boolean" ||
    typeof value === "bigint"
  )
    return String(value);
  return JSON.stringify(value);
}

/**
 * Parse an AGS4-shaped raw value into its canonical JS value (empty /
 * unparseable → null). datetime/date/time come back as the canonical **string**
 * (engine shape; `new Date(s)` if you want a Date). Non-string input is
 * stringified first (matches the Python wrapper).
 *
 * @param raw - The raw cell value; non-string input is stringified, and `null`/`undefined` short-circuit to `null`.
 * @param agsType - The AGS4 spec type code governing how `raw` is interpreted.
 * @returns The canonical value: a number (integer/decimal), boolean (YN), string (text/enum and the datetime/date/time strings), or `null`.
 */
export function parseValue(raw: unknown, agsType: string): AgsValue {
  const s = raw === null || raw === undefined ? null : scalarString(raw);
  return nativeParseValue(s, agsType) as AgsValue;
}
